src/EditorFeatures/TestUtilities/Threading/StaTaskScheduler.cs

```diff
@@ -55,6 +58,34 @@ public StaTaskScheduler()
                 DispatcherSynchronizationContext = synchronizationContext;
             }
 
+            // We've created an STA thread, which has some extra requirements for COM Runtime
+            // Callable Wrappers (RCWs). If any COM object is created on the STA thread, calls to that
+            // object must be made from that thread; when the RCW is no longer being used by any
+            // managed code, the RCW is put into the finalizer queue, but to actually finalize it
+            // it has to marshal to the STA thread to do the work. This means that in order to safely
+            // clean up any RCWs, we need to ensure that the thread is pumping past the point of
+            // all RCWs being finalized
+            //
+            // This constraint is particularly problematic if our tests are running in an AppDomain:
+            // when the AppDomain is unloaded, any threads (including our STA thread) are going to be
+            // aborted. Once the thread and AppDomain is being torn down, the CLR is going to try cleaning up
+            // any RCWs associated them, because if the thread is gone for good there's no way
+            // it could ever clean anything further up. The code there waits for the finalizer queue
+            // -- but the finalizer queue might be already trying to clean up an RCW, which is marshaling
+            // to the STA thread. This could then deadlock.
+            //
+            // The suggested workaround from the CLR team is to do an explicit GC.Collect and
+            // WaitForPendingFinalizers before we let the AppDomain shut down. The belief is subscribing
+            // to the Unload is a reasonable place to do it.
+            if (Interlocked.CompareExchange(ref s_SubscribedToAppDomainUnload, 1, 0) == 0)
+            {
+                AppDomain.CurrentDomain.DomainUnload += (sender, e) =>
```

Comment:
Does this event happen before or after `Thread.Abort` is called on our STA thread?

Reply:
From the mail thread with the CLR team, before.

Reply:
Hmm ... if that's the case I wonder if we should call `Dispatcher.Shutdown` here as well. That would seemingly solve the unload event race that's displayed below. That's a separate issue though. 

---
In reply to: [269633055](https://github.com/dotnet/roslyn/pull/34498#discussion_r269633055) [](ancestors = 269633055)

Reply:
That would work. I guess there's also good arguments here that we should just be making this thread a test fixture so it's properly shut down when we're done using it.

Reply:
I looked into that a bit. The setup works and we have a definite time in which to do shutdown operations. At the same time though:

1. It still requires passing `static` data. There is no way for the test fixture and test runner to communicate via the xunit APIs. That means mutable `static` is the only way. 
1. It requires a lot of source annotations. Including duplicating definitions between all our test assemblies. That's annoying but not a deal breaker. 

That's why I haven't moved forward on it.

Reply:
Ah, right forgot about 1. I think the "usual" asumption is instead we'd switch our tests to regular Facts and just have all our Wpf ones do something like _staThread.RunOnThread() and then pass the lambda of the actual test contents. Or do an await SwitchToStaThread() pattern not unlike JTF.